src/components/lib/cos_kernel_api.c

```diff
@@ -456,6 +521,25 @@ __page_bump_valloc(struct cos_compinfo *ci, size_t sz)
 	return __page_bump_mem_alloc(ci, &ci->vas_frontier, &ci->vasrange_frontier, sz);
 }
 
+static vaddr_t
+__superpage_bump_valloc(struct cos_compinfo *ci, size_t sz) {
+	vaddr_t inc, ret = 0;
+	vaddr_t *ptr, *frontier;
+
+	ptr      = &ci->vas_frontier;
+	frontier = &ci->vasrange_frontier;
+
+	/* RSK -- expand heap frontier if necessary */
+	if (*ptr + sz > *frontier) {
+		inc = round_up_to_pgd_page(*ptr + sz) - *frontier;
+		ret = ps_faa(frontier, inc);
+	}
+	inc = round_up_to_pgd_page(*ptr + sz) - *ptr;
+	ret = ps_faa(ptr, inc);
```

Comment:
this can fail due a race.  Should likely have a loop to re-expand the vas frontier in case of this failure.